cpp/src/arrow/util/value_parsing.h

```diff
@@ -484,6 +455,14 @@ static inline bool ParseHH_MM_SS(const char* s, std::chrono::duration<ts_type>*
 
 }  // namespace detail
 
+/// \brief Attempt to convert a string to the primitive type corresponding to
+/// an Arrow data type
+template <typename T, typename ParseContext = void>
+inline bool ParseValue(const char* s, size_t length, typename T::c_type* out,
+                       const ParseContext* ctx = NULLPTR) {
+  return detail::StringConverter<T>::Convert(s, length, out);
+}
+
```

Comment:
This declaration (and the specialization below) make the requirements on ParseContext extremely unclear. It's optional and may be of any type... except when parsing timestamps. Instead it seems better to make ParseContext required for all types and empty for the majority: 
```suggestion
template <typename T>
struct ParseContext {};

template <>
struct ParseContext<TimestampType> {
  TimeUnit::type unit;
};

/// \brief Attempt to convert a string to the primitive type corresponding to
/// an Arrow data type
template <typename T, typename Converter = detail::StringConverter<T>>
inline bool ParseValue(const ParseContext<T>& ctx, util::string_view s, typename T::c_type* out) {
  return Converter{}::Convert(s.data(), s.size(), out);
}

template <>
inline bool ParseValue<TimestampType>(const ParseContext<TimestampType>& ctx,
    util::string_view s, typename T::c_type* out) {
  return ParseTimestampISO8601(s.data(), s.size(), ctx.unit, out);
}
```

This makes the usage of a parse context clearer and does not add significantly to boilerplate at any call site:
```c++
ParseValue<Int8Type>({}, "42", &value)
ParseValue<TimestampType>({TimeUnit::NANO}, ts_str, &ts_value)
```

Bonus: the out argument is the last argument again.

Reply:
I still really don't like this because you're pushing requirements from a special case onto the general case

Reply:
It's confusing when signatures for generics like this differ significantly, so yes: as I wrote it the possibility of a non-empty context does push the requirement of a context argument onto overloads of `ParseValue` which are context free.

If you feel `ParseValue<TimestampType>()` has requirements which are sufficiently different from `ParseValue<T>()` that you think the signatures *should* be different, then why should they be in the same overload family? What value does `ParseValue<TimestampType>()` provide over `ParseTimestampISO8601()`?

Reply:
It doesn't provide much value. I was trying to address @pitrou's comment about parsers that may require some state initialization like a locale and I used Timestamp as an example